### PR TITLE
feat(PRT): introduce Particle Tracking (PRT) model

### PR DESCRIPTION
**Pull request by wpbonelli**
#### Status

At this time PRT supports particle tracking on all DIS and some but not all DISV grids

Outstanding issues include

- DISV cells containing adjacent parallel faces are not properly handled and may cause crashes, this includes quad-refined grids as well as DISV grids generated by FloPy's `voronoi` module

Support is planned but not yet implemented for

- configurable particle mass
- local mass conservation &mdash; conserved between cells but not yet guaranteed at the subcell level
- exchanging particles between models
- DISU grids (longer term)
- solving models in parallel (longer term)

#### Integration

When this merges a preliminary version of PRT will become available on the nightly build but will be excluded from official releases.

#### Miscellaneous info

Docs/dfns and tests make up about half the changeset, the rest is fortran source.

PRT's primary author is @aprovost-usgs. Some code is adapted from MODPATH and originally due to @dwpolloc-usgs.

### CODE REVIEW

**Comment by wpbonelli** [comment]
Squashed the branch to make rebasing with [IDM](https://github.com/MODFLOW-USGS/modflow6/commit/78bfda99381e85b2f091cd1299fcab6f251db688) easier. Commit history is backed up [on my fork](https://github.com/wpbonelli/modflow6/commits/PRT-wip/) if needed. Recent changes are mostly non-functional and include

* refactor tracking method hierarchy
* refactor cell/cell definition hierarchy, factor out base cell type
* refactor prp: simplify time offset support, rename referencetime -> releasetime, check release elevation, cleanup particle init/release logic
* refactor TrackModule to support arbitrary # of track files
* remove trackingmodel, extend explicitmodel directly
* don't set conc on gwfprt exchange until implemented/tested
* deduplicate logic in Method types, use dis%get_polyverts()
* move coord transform routines to GeomUtil, remove UtilMisc
* use ExpandArray interface, remove allocate_as_needed utils
* use arrays for canonical velocity components, vertices, etc
* introduce mod_offset function, remove *_wrap routines
* remove rotate function, replace rotate() with matmuls
* remove some unused utility routines
* minor cleanup in MIP pkg
* ifort compatibility tweaks
* add scipy to environment.yml
* rebase with latest IDM changes
* add unit tests for math/geom utils
* add integration tests: drape, voronoi grid

**Comment by wpbonelli** [comment]
@mjr-deltares

Do you have any ideas on this? https://github.com/MODFLOW-USGS/modflow6/actions/runs/7343758410/job/19994783536?pr=1389#step:15:392

I'm stumped. PRT doesn't touch PETSc and it doesn't reproduce on ubuntu 22.04 locally. Tried clearing caches, no luck

File: src/Solution/ParticleTracker/MethodCellTernary.f90 [hunk 1]
```diff
@@ -17,13 +18,13 @@ module MethodCellTernaryModule
...
     private
-    double precision :: x_vert(99), y_vert(99) !< cell vertex coordinates (kluge 99, todo: allocate as needed)
-    double precision :: xctr, yctr !< cell center coordinates
-    double precision :: vx_vert_polygon(99), vy_vert_polygon(99) !< cell vertex velocities
-    double precision :: vxctr, vyctr !< cell center velocities
-    double precision :: ztop, zbot !< cell top and bottom elevations
-    double precision :: dz !< cell thickness
-    double precision :: vztop, vzbot !< cell top and bottom velocities
+    real(DP), allocatable :: x_vert(:), y_vert(:) !< cell vertex coordinates
+    real(DP), allocatable :: vx_vert_polygon(:), vy_vert_polygon(:) !< cell vertex velocities
```
Comment (jdhughes-usgs) [inline src/Solution/ParticleTracker/MethodCellTernary.f90:22]: style-wise I think we want variables on their own lines.

Comment (wpbonelli) [reply src/Solution/ParticleTracker/MethodCellTernary.f90:22]: ok. I will go back and update here and elsewhere

**Review by langevin-usgs — APPROVED** [review]
Looks great.

File: doc/mf6io/mf6ivar/tex/prt-nam-desc.tex [hunk 2]
```diff
@@ -0,0 +1,25 @@
+% DO NOT MODIFY THIS FILE DIRECTLY.  IT IS CREATED BY mf6ivar.py 
+
+\item \textbf{Block: OPTIONS}
+
+\begin{description}
+\item \texttt{list}---is name of the listing file to create for this PRT model.  If not specified, then the name of the list file will be the basename of the PRT model name file and the '.lst' extension.  For example, if the PRT name file is called ``my.model.nam'' then the list file will be called ``my.model.lst''.
```
[the remark this answers is not in the document]
Comment (wpbonelli) [reply doc/mf6io/mf6ivar/tex/prt-nam-desc.tex:6]: Ok. Should we update this in `*-nam.dfn` for the other models too?

[the remark this answers is not in the document]
Comment (langevin-usgs) [reply doc/mf6io/mf6ivar/tex/prt-nam-desc.tex:6]: If it's like that elsewhere, then maybe we are okay for now.

**Pull request merged**